---
title: "Build Scaffold"
output: 
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
    highlight: pygments
author: "SÃ¸ren Helweg Dam"
date: "Last updated: `r format(Sys.Date(), '%Y.%m.%d')`"
vignette: >
  %\VignetteIndexEntry{Build Scaffold}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

# Quick overview


```{r tldr}
library(spaceRAT)

# Load expression data
data("DMAP_exprs", package="spaceRATScaffolds")
# Load phenotype data
data("DMAP_pData", package="spaceRATScaffolds")
# Specify colname with phenotype information
colname <- "cell_types"

# Build scaffold
scaffold <- buildScaffold(
    DMAP_exprs, 
    pheno = DMAP_pData,
    data = "exprs",
    colname = colname,
    rank_scale = TRUE
)

# Plot scaffold
plotScaffold(scaffold, title = "Scaffold plot")
```

See `?buildScaffold` for additional parameter settings.


# Introduction

This vignette will describe how to create a scaffold space
that samples can later be projected into.
You will need:

- Counts or expression matrix 
- Phenotype data that describes the desired phenotype of each
column of the matrix.

*Counts/expression data*

The matrix may be formatted as a `matrix`, `data.frame`,
or `r  BiocStyle::Biocpkg("SummarizedExperiment")`.
Column names are sample names. Row names are gene names,
which can be Ensembl Gene ID, HGNC Symbol, Entrez Gene ID Ensembl,
Transcript ID, or Refseq mRNA. All gene ID will be automatically converted to
Ensembl Gene ID (Unless otherwise specified by the user).
Counts of several transcript ID corresponding to same gene
will be added and recorded as the counts of the gene.

Below is a valid (log-transformed) count matrix:
```{r expression}
data("DMAP_exprs", package="spaceRATScaffolds")
DMAP_exprs[1:5,1:5]
```


*Phenotype*

Phenotype table should have sample names as row names, such that they are
identical to the column names of the data matrix,
as well as at least one column for cell types, or any sample feature
suitable for grouping cells and perform differential expression analysis.
For example:

```{r pdata}
data("DMAP_pData", package="spaceRATScaffolds")
DMAP_pData[1:5,,drop=FALSE]
```


# Load data

The first step is to load in the counts/expression data.

```{r load}
library(spaceRAT)

if(FALSE){
    # Data in csv
    counts_scaffold <- read.csv("path/to/counts_data.csv")
    pheno_scaffold <- read.csv("path/to/pheno.csv")
    # Data in SummarizedExperiment stored as .rds
    se <- readRDS("path/to/data.RDS")
}

# Column name of interest
colname <- "cell_types"
```
Here, `"cell_types"` is the name of a column of `pData_dmap`,
according to which cells are grouped and labeled.
Differential expression analysis will also be performed using this
column of phenotype as independent variables. 

# Build scaffold

From there, the scaffold can be built with `buildScaffold`.

```{r build}
scaffold <- buildScaffold(
    DMAP_exprs,
    pheno = DMAP_pData,
    data = "exprs",
    colname = colname,
    add_umap = FALSE, # Set to TRUE if you also want UMAP space in the scaffold
    rank_scale = TRUE # This will scale the ranks to account for zeros
)


str(scaffold)
```

If you want to add a UMAP and run into irlba/Matrix issues, it can be solved
by reinstalling both packages from source:

```{r, eval=FALSE}
install.packages("Matrix", type = "source")
install.packages("irlba", type = "source")
```

`buildScaffold()` performs the following steps:

- Preprocesses the counts/expression matrix: It removes genes with total counts
less than 10 and replaces `NA` with 0.
- Differential expression analysis to select 
differentially expressed genes (DE genes).
- Subsets the scaffold dataset to contain only the DE genes. 
    - (This can be disabled with `subset_deg = FALSE`)
- Ranks the genes within each sample
- Finally, it performs principle component analysis (PCA) or UMAP using ranks.

By default `buildScaffold()` takes in a counts matrix.
It can also handle expression data if `data = "exprs"` is specified. 

The resulting scaffold has 8-9 slots:

- `DEgene`: The differentially expressed genes, determined using
`r  BiocStyle::Biocpkg("limma")`.
- `label`: The phenotype labels from the provided pheno data.
- `pca`: The PCA space
- `umap`: The UMAP space
- `annotation`: The annotation used for the scaffold
- `rank`: The ranks computed for the scaffold 
(in case the PCA needs recompution)
- `pca_scale`: Whether the PCA was scaled
- `rank_scale`: Whether ranks were scaled to the maximum rank 
(i.e., the lowest expression value)
- `spaceRATVersion`: The version of spaceRAT the scaffold was built with.



The scaffold space can be plotted with `plotScaffold`:

Here, you specify if the PCA or UMAP should be used,
the PCs to plot, and the plotmode (`dot` or `tiny_label`)

```{r plotscaffold}
plotScaffold(
    scaffold,
    title = "Scaffold plot",
    dimred = "PCA",
    plot_mode = "dot",
    dims = c(1,2))
```

You can also turn on the "tiny_label" mode by specifying `plot_mode`.
You can also visualize other principle components by specifying `dims`.  
```{r}
plotScaffold(
    scaffold,
    dimred = "PCA",
    plot_mode = "tiny_label",
    dims = c(3,4),
    title = "Scaffold Plot")
```
## Loading plot

To make sense of scaffold space, `loadingPlot()` can visualize the genes
that contribute most to the selected principle components.
It returns a data frame containing
the loading scores by setting `df_only = TRUE`.
```{r, eval=FALSE}
loadingPlot(
    scaffold,
    df_only = FALSE,
    dims = c(1,2),
    num_genes = 3,
    gene_name = "hgnc_symbol"
    )
```


## Building from a `r  BiocStyle::Biocpkg("SummarizedExperiment")`

If you have a `r  BiocStyle::Biocpkg("SummarizedExperiment")` object,
the process is very similar.

```{r SE}
# Prepare data
se <- SummarizedExperiment::SummarizedExperiment(
    assays = S4Vectors::SimpleList("exprs" = DMAP_exprs),
    colData = S4Vectors::DataFrame(DMAP_pData)
)
colname <- "cell_types"

# Build scaffold

scaffold <- buildScaffold(
    se,
    assay = "exprs",
    data = "exprs",
    colname = colname,
    rank_scale = TRUE
)

# Plot scaffold
plotScaffold(scaffold, title = "Scaffold plot")
```

See `?buildScaffold` for additional parameter settings.

# Session Info

```{r session info}
sessionInfo()
```
