---
title: "Build Scaffold"
output: 
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
    highlight: pygments
author: "SÃ¸ren Helweg Dam"
date: "Last updated: `r format(Sys.Date(), '%Y.%m.%d')`"
vignette: >
  %\VignetteIndexEntry{Build Scaffold}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

# Quick overview


```{r tldr}
library(spaceRAT)

# Load expression data
data("exprs_dmap", package="spaceRATScaffolds")
# Load phenotype data
data("pData_dmap", package="spaceRATScaffolds")
# Specify colname with phenotype information
colname <- "cell_types"

# Build scaffold
scaffold <- buildScaffold(
    exprs_dmap, 
    pheno_scaffold = pData_dmap,
    colname = colname,
    title = "DMAP Scaffold"
)

# Plot scaffold
plotScaffold(scaffold, title = "Scaffold plot")
```

See `?buildScaffold` for additional parameter settings.


# Introduction

This vignette will describe how to create a scaffold space
that samples can later be projected into.
You will need:
- Raw count or expression matrix 
- Phenotype data that describes the desired phenotype of each
column of the matrix.

*Count/expression data*

The matrix may be formatted as a `matrix`, `data.frame`,
or `r  BiocStyle::Biocpkg("SummarizedExperiment")`.
Column names are sample names. Row names are gene names,
which can be Ensembl Gene ID, HGNC Symbol, Entrez Gene ID Ensembl,
Transcript ID or Refseq mRNA. All gene ID will be automatically converted to
Ensembl Gene ID. Counts of several transcript ID corresponding to same gene
will be added and recorded as the counts of the gene.

Below is a valid (log-transformed) count matrix:
```{r expression}
data("exprs_dmap", package="spaceRATScaffolds")
exprs_dmap[1:5,1:5]
```


*Phenotype*

Phenotype table should have row names as sample names identical to count matrix,
as well as at least one column for cell types, or any sample feature
suitable for grouping cells and perform differential expression analysis.
For example:

```{r pdata}
data("pData_dmap", package="spaceRATScaffolds")
pData_dmap[1:5,,drop=FALSE]
```


# Load data

The first step is to load in the raw data.

```{r load}
library(spaceRAT)

if(FALSE){
    # Data in csv
    counts_scaffold <- read.csv("path/to/raw_data.csv")
    pheno_scaffold <- read.csv("path/to/pheno.csv")
    # Data in SummarizedExperiment stored as .rds
    se <- readRDS("path/to/data.RDS")
}

# Column name of interest
colname <- "cell_types"
```
Here, `"cell_types"` is the name of a column of `pData_dmap`,
according to which cells are grouped and labeled.
Differential expression analysis will also be performed using this
column of phenotype as independent variables. 

# Build scaffold

From there, the scaffold can be built with `buildScaffold`.

```{r build}
scaffold <- buildScaffold(
    exprs_dmap,
    pheno_scaffold = pData_dmap,
    colname = colname,
    title = "DMAP Scaffold"
)


str(scaffold)
```

`buildScaffold()` performs the following steps:

- Preprocesses the count matrix: It removes genes with total counts
less than 10 and replaces `NA` by 0.
- Differential expression analysis to select 
differentially expressed genes (DE genes).
- Subsets the scaffold dataset to contain only the DE genes.
- Ranks the genes within each sample
- Finally, it performs principle component analysis (PCA) or UMAP using ranks.

By default `buildScaffold()` takes in logged count matrix.
It can also handle raw count data if `data="raw"` is specified. 

The resulting scaffold has five slots:

- `DEgene`: The differentially expressed genes, determined using
`r  BiocStyle::Biocpkg("limma")`.
- `label`: The phenotype labels from the provided pheno data.
- `model`: The `PCA` or `UMAP` dimensionality reduction model.
- `dims`: The principal components to use,
- `plot_mode`: The plotting type specified when building the scaffold.
Can be either `"dot"` or `"tiny_label"`.


The scaffold space can be plotted with `plotScaffold`:

```{r plotscaffold}
plotScaffold(scaffold, title = "Scaffold plot")
```

You can also turn on the "tiny_label" mode by specifying `plot_mode`.
You can also visualize other principle components by specifying `dims`.  
```{r}
scaffold <- buildScaffold("DMAP_scaffold",plot_mode = "tiny_label",dims=c(3,4))
plotScaffold(scaffold, "Scaffold Plot")
```

To make sense of scaffold space, `loadingPlot()` can visualize the genes
that contribute most to the
selected principle components. It returns a data frame containing
the loading scores as well by setting `df_only = TRUE`.
```{r}
loadingPlot(
    scaffold,
    df_only = FALSE
    )
```


## Building from a `r  BiocStyle::Biocpkg("SummarizedExperiment")`

If you have a `r  BiocStyle::Biocpkg("SummarizedExperiment")` object,
the process is very similar.

```{r SE}
# Prepare data
se <- SummarizedExperiment::SummarizedExperiment(
    assays = S4Vectors::SimpleList("exprs" = exprs_dmap),
    colData = S4Vectors::DataFrame(pData_dmap)
)
colname <- "cell_types"

# Build scaffold

scaffold <- buildScaffold(
    se,
    assay = "exprs",
    colname = colname,
    title = "DMAP Scaffold"
)

# Plot scaffold
plotScaffold(scaffold, title = "Scaffold plot")
```

See `?buildScaffold` for additional parameter settings.

# Session Info

```{r session info}
sessionInfo()
```
